File: pkg/datasource/volume.go
package datasource

import (
	"context"
	"fmt"
	"os/exec"
	"strconv"
	"strings"
	"time"
)

type VolumeDataSource struct {
	callbacks []func(interface{})
	currentValue VolumeData
}

func NewVolumeDataSource() *VolumeDataSource {
	return &VolumeDataSource{
		callbacks: make([]func(interface{}), 0),
		currentValue: VolumeData{
			Volume:       0,
			VolumeString: "🔇 --",
			IsMuted:      false,
		},
	}
}

func (ds *VolumeDataSource) Start(ctx context.Context) {
	go func() {
		ticker := time.NewTicker(1 * time.Second)
		defer ticker.Stop()
		
		// Get initial value
		ds.updateVolume()
		
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				ds.updateVolume()
			}
		}
	}()
}

func (ds *VolumeDataSource) Subscribe(callback func(interface{})) {
	ds.callbacks = append(ds.callbacks, callback)
}

func (ds *VolumeDataSource) GetCurrentValue() interface{} {
	return ds.currentValue
}

func (ds *VolumeDataSource) ToggleMute() {
	toggleMute()
	ds.updateVolume()
}

func (ds *VolumeDataSource) ChangeVolume(direction int) {
	changeVolume(direction)
	ds.updateVolume()
}

func (ds *VolumeDataSource) updateVolume() {
	volume, isMuted := getVolumeInfo()
	
	var volumeString string
	if isMuted {
		volumeString = "🔇 --%"
	} else {
		var icon string
		if volume == 0 {
			icon = "🔇"
		} else if volume < 33 {
			icon = "🔈"
		} else if volume < 66 {
			icon = "🔉"
		} else {
			icon = "🔊"
		}
		volumeString = fmt.Sprintf("%s %d%%", icon, volume)
	}
	
	newValue := VolumeData{
		Volume:       volume,
		VolumeString: volumeString,
		IsMuted:      isMuted,
	}
	
	ds.currentValue = newValue
	
	// Notify all subscribers
	for _, callback := range ds.callbacks {
		callback(newValue)
	}
}

func getVolumeInfo() (int, bool) {
	// Try PulseAudio first
	if volume, isMuted, err := getPulseAudioVolume(); err == nil {
		return volume, isMuted
	}
	
	// Try ALSA as fallback
	if volume, isMuted, err := getALSAVolume(); err == nil {
		return volume, isMuted
	}
	
	return 0, false
}

func getPulseAudioVolume() (int, bool, error) {
	// Get volume
	cmd := exec.Command("pactl", "get-sink-volume", "@DEFAULT_SINK@")
	output, err := cmd.Output()
	if err != nil {
		return 0, false, err
	}
	
	// Parse volume from output like "Volume: front-left: 32768 /  50% / -18.06 dB"
	lines := strings.Split(string(output), "\n")
	for _, line := range lines {
		if strings.Contains(line, "Volume:") {
			parts := strings.Split(line, "/")
			if len(parts) >= 2 {
				volumePart := strings.TrimSpace(parts[1])
				volumePart = strings.TrimSuffix(volumePart, "%")
				if volume, err := strconv.Atoi(volumePart); err == nil {
					// Check if muted
					cmd := exec.Command("pactl", "get-sink-mute", "@DEFAULT_SINK@")
					muteOutput, err := cmd.Output()
					if err != nil {
						return volume, false, nil
					}
					
					isMuted := strings.Contains(string(muteOutput), "Mute: yes")
					return volume, isMuted, nil
				}
			}
		}
	}
	
	return 0, false, fmt.Errorf("could not parse volume")
}

func getALSAVolume() (int, bool, error) {
	// Try to get volume from ALSA
	cmd := exec.Command("amixer", "get", "Master")
	output, err := cmd.Output()
	if err != nil {
		return 0, false, err
	}
	
	lines := strings.Split(string(output), "\n")
	for _, line := range lines {
		if strings.Contains(line, "[") && strings.Contains(line, "%") {
			// Parse line like "  Front Left: Playbook 32768 [50%] [18.06dB] [on]"
			start := strings.Index(line, "[")
			end := strings.Index(line, "%]")
			if start != -1 && end != -1 {
				volumeStr := line[start+1 : end]
				if volume, err := strconv.Atoi(volumeStr); err == nil {
					// Check if muted (look for [off])
					isMuted := strings.Contains(line, "[off]")
					return volume, isMuted, nil
				}
			}
		}
	}
	
	return 0, false, fmt.Errorf("could not parse ALSA volume")
}

func toggleMute() {
	// Try PulseAudio first
	cmd := exec.Command("pactl", "set-sink-mute", "@DEFAULT_SINK@", "toggle")
	if err := cmd.Run(); err != nil {
		// Try ALSA as fallback
		cmd = exec.Command("amixer", "set", "Master", "toggle")
		cmd.Run()
	}
}

func changeVolume(direction int) {
	volumeChange := direction * 5 // Change by 5%
	
	// Try PulseAudio first
	var sign string
	if volumeChange > 0 {
		sign = "+"
	} else {
		sign = ""
	}
	
	cmd := exec.Command("pactl", "set-sink-volume", "@DEFAULT_SINK@", fmt.Sprintf("%s%d%%", sign, volumeChange))
	if err := cmd.Run(); err != nil {
		// Try ALSA as fallback
		var alsaDirection string
		if volumeChange > 0 {
			alsaDirection = fmt.Sprintf("%d%%+", volumeChange)
		} else {
			alsaDirection = fmt.Sprintf("%d%%-", -volumeChange)
		}
		cmd = exec.Command("amixer", "set", "Master", alsaDirection)
		cmd.Run()
	}
}

File: src/sources/volume.rs
//! Audio volume via PulseAudio (pactl) with ALSA (amixer) fallback.
//! Port of pkg/datasource/volume.go.

use std::process::Command;

#[allow(dead_code)] // volume/is_muted mirror the Go model
#[derive(Debug, Clone)]
pub struct VolumeData {
    pub volume: i32,
    pub string: String,
    pub is_muted: bool,
}

impl Default for VolumeData {
    fn default() -> Self {
        VolumeData {
            volume: 0,
            string: "--".to_string(),
            is_muted: false,
        }
    }
}

pub fn update_volume() -> VolumeData {
    let (volume, is_muted) = get_volume_info();
    VolumeData {
        volume,
        string: format_volume(volume, is_muted),
        is_muted,
    }
}

fn format_volume(volume: i32, is_muted: bool) -> String {
    if is_muted {
        return "--%".to_string();
    }
    // the speaker icon is now a separate SVG widget (chosen by level in the module)
    format!("{}%", volume)
}

fn get_volume_info() -> (i32, bool) {
    if let Some(v) = get_pulseaudio_volume() {
        return v;
    }
    if let Some(v) = get_alsa_volume() {
        return v;
    }
    (0, false)
}

fn get_pulseaudio_volume() -> Option<(i32, bool)> {
    let output = Command::new("pactl")
        .args(["get-sink-volume", "@DEFAULT_SINK@"])
        .output()
        .ok()?;
    if !output.status.success() {
        return None;
    }
    let text = String::from_utf8_lossy(&output.stdout);
    for line in text.lines() {
        if line.contains("Volume:") {
            let parts: Vec<&str> = line.split('/').collect();
            if parts.len() >= 2 {
                let vp = parts[1].trim().trim_end_matches('%').trim();
                if let Ok(volume) = vp.parse::<i32>() {
                    let mute = Command::new("pactl")
                        .args(["get-sink-mute", "@DEFAULT_SINK@"])
                        .output();
                    let is_muted = match mute {
                        Ok(m) => String::from_utf8_lossy(&m.stdout).contains("Mute: yes"),
                        Err(_) => false,
                    };
                    return Some((volume, is_muted));
                }
            }
        }
    }
    None
}

fn get_alsa_volume() -> Option<(i32, bool)> {
    let output = Command::new("amixer")
        .args(["get", "Master"])
        .output()
        .ok()?;
    if !output.status.success() {
        return None;
    }
    let text = String::from_utf8_lossy(&output.stdout);
    for line in text.lines() {
        if line.contains('[') && line.contains('%') {
            let start = line.find('[');
            let end = line.find("%]");
            if let (Some(s), Some(e)) = (start, end) {
                if let Ok(volume) = line[s + 1..e].parse::<i32>() {
                    let is_muted = line.contains("[off]");
                    return Some((volume, is_muted));
                }
            }
        }
    }
    None
}

pub fn toggle_mute() {
    let res = Command::new("pactl")
        .args(["set-sink-mute", "@DEFAULT_SINK@", "toggle"])
        .status();
    if res.map(|s| !s.success()).unwrap_or(true) {
        let _ = Command::new("amixer")
            .args(["set", "Master", "toggle"])
            .status();
    }
}

/// direction: +1 to raise, -1 to lower (by 5%).
pub fn change_volume(direction: i32) {
    let change = direction * 5;
    let sign = if change > 0 { "+" } else { "" };
    let pulse = Command::new("pactl")
        .args([
            "set-sink-volume",
            "@DEFAULT_SINK@",
            &format!("{}{}%", sign, change),
        ])
        .status();
    if pulse.map(|s| !s.success()).unwrap_or(true) {
        let alsa_dir = if change > 0 {
            format!("{}%+", change)
        } else {
            format!("{}%-", -change)
        };
        let _ = Command::new("amixer")
            .args(["set", "Master", &alsa_dir])
            .status();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn formats_by_level() {
        assert_eq!(format_volume(0, false), "0%");
        assert_eq!(format_volume(20, false), "20%");
        assert_eq!(format_volume(40, false), "40%");
        assert_eq!(format_volume(80, false), "80%");
        assert_eq!(format_volume(50, true), "--%");
    }
}
